### ISSUE TITLE
Spurious and incorrect "irrefutable `if let` pattern" warning in the presence of or-patterns

### ISSUE THREAD

**Issue by bstrie** [rust-lang/rust#102411]
The following program:
```rust
fn main() {
    let x: Result<i32, i32> = Ok(42);
    if let Ok(_) | Err(_) = x {}
}
```
...compiles successfully, but produces a warning:
```
warning: irrefutable `if let` pattern
 --> src/main.rs:3:8
  |
3 |     if let Ok(_) | Err(_) = x {}
  |        ^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(irrefutable_let_patterns)]` on by default
  = note: this pattern will always match, so the `if let` is useless
  = help: consider replacing the `if let` with a `let`
```
..however, following the warning's advice to replace the `if let` with a `let`, resulting in the following program:
```rust
fn main() {
    let x: Result<i32, i32> = Ok(42);
    let Ok(_) | Err(_) = x;
}
```
..results in compilation failing with the following error:
```
error: top-level or-patterns are not allowed in `let` bindings
 --> src/main.rs:4:9
  |
3 |     let Ok(_) | Err(_) = x;
  |         ^^^^^^^^^^^^^^ help: wrap the pattern in parentheses: `(Ok(_) | Err(_))`
```
In the future, if/when top-level or-patterns are allowed in `let`, then this warning would be correct. However, for the moment, this warning is mistaken, and should either not be emitted or should suggest the version wrapped in parentheses.

**Comment by scottmcm** [comment]
There is an alternative; you wrap the pattern in parens, as the `help` says:
```rust
fn main() {
    let x: Result<i32, i32> = Ok(42);
    let (Ok(_) | Err(_)) = x;
}
```
<https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3f741264b0440f42e374a105993c0df8>

**Comment by bstrie** [comment]
I've updated the issue, still a bit annoying though.

**Comment by shanecelis** [comment]
I had this issue for an enum with only one variant. When I tried to make a minimal example, I could not provoke it. The weird thing was, I changed it to a `let` no `if` and it still complained. And I even added the `allow(irrefutable_let_patterns)` and it would not stop showing the warning on `cargo check` (not clippy).

```rust
//! Failed reproduction of issue
pub enum Layer {
    Liquid { id: usize, volume: f32 },
}

/// The warning this used to provoke in my code that I could not force it to
/// issue in this minimal example with rustc v1.78.0 (9b00956e5 2024-04-29)
///
/// ```
/// warning: irrefutable `if let` pattern
///  --> src/level.rs:67:23
///   |
///67 |                   let Layer::Liquid { id, .. } = l;
///   |  _______________________^
///68 | |                 *id
///   | |____^
///   |
///   = note: this pattern will always match, so the `if let` is useless
///   = help: consider replacing the `if let` with a `let`
///   = note: `#[warn(irrefutable_let_patterns)]` on by default
/// ```
// #[allow(irrefutable_let_patterns)]
fn f(l: &Layer) -> Option<usize> {
    let Layer::Liquid { id, .. } = l;
    Some(*id)
}

fn main() {
    let l = Layer::Liquid { id: 0, volume: 0.0 };
    println!("{:?}", f(&l));
}
```

**Comment by kands-code** [comment]
I've seen a piece of code where the compiler also gives this warning, but I don't think there's a cleaner way to write it, and this code is from the [typst](https://github.com/typst/typst/blob/e0d809680aed778443a4797bc044376dbe15347e/crates/typst/src/foundations/int.rs#L405).

```rust
macro_rules! unsigned_int {
    ($($ty:ty)*) => {
        $(cast! {
            $ty,
            self => if let Ok(int) = i64::try_from(self) { // => warning here
                Value::Int(int)
            } else {
                // Some u64 are too large to be cast as i64
                // In that case, we accept that there may be a
                // precision loss, and use a floating point number
                Value::Float(self as _)
            },
            v: i64 => v.try_into().map_err(|_| {
                if v < 0 {
                    "number must be at least zero"
                } else {
                    "number too large"
                }
            })?,
        })*
    }
}
```

The warning is:

```plaintext
warning: irrefutable `if let` pattern
   --> crates/typst/src/foundations/int.rs:405:24
    |
405 |             self => if let Ok(int) = i64::try_from(self) {
    |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
425 | unsigned_int! { u8 u16 u32 u64 usize }
    | -------------------------------------- in this macro invocation
    |
    = note: this pattern will always match, so the `if let` is useless
    = help: consider replacing the `if let` with a `let`
    = note: `#[warn(irrefutable_let_patterns)]` on by default
    = note: this warning originates in the macro `unsigned_int` (in Nightly builds, run with -Z macro-backtrace for more info)
```
